contexts/did-core-latest.jsonld

```diff
@@ -0,0 +1,16 @@
+{
```

Comment:
Don't put this file here... it exists in did-core, should probably stay there -- https://github.com/w3c/did-core/tree/master/contexts

We may move it over in the future, but for now, let's not change too many structural things. Remove this file, put it in did-core via another PR.

Reply:
I'm not sure I agree with this. The purpose of this registry is to consolidate machine testability and human documentation.... we tried having this stuff be defined in did core, and it did not work.

I'd like to define JSON Schema + JSON-LD + Human documentation in one place and it seems like that will have to be the registry given the current structure.

Reply:
> The purpose of this registry is to consolidate machine testability and human documentation.

The purpose of the registry, based on the DID F2F meeting, is to enable global interoperability. Human documentation was one of the only things we really had agreement on. Machine testability is desirable, I agree with you, but others in the group might not see it that way and we can just move forward assuming that.

Part of being an Editor is to think about if anyone is going to object to this change w/o the appropriate change happening in the DID Core spec, which is going to have to involve raising and discussing this with the group and other implementers. I'm trying to get this PR in quickly, but duplicating some did-core-latest.jsonld context that does not have all the values that an implementer would expect in it is problematic... because if people start using the did-core-latest.jsonld context that is a part of this PR in its current form, now the context is split between two repos where it's not clear which repo context you're supposed to use as an implementer -- this makes the problem you highlighted and are trying to fix worse... but more importantly

> we tried having this stuff be defined in did core, and it did not work.

What didn't work? The context was out of date? The context was out of date because we didn't have consensus on what to change it to...

You also seem to be merging the did/v1 context with the registry context... there is a potential future here where there may be two contexts: the did core one, and a dated extension registry one. Barreling forward and trying to merge them will result in disaster without a solid caching/upgrade strategy. What is that caching/upgrade strategy? Putting things in the did-core-latest.jsonld context that don't have consensus, or removing things w/o consensus is also going to result in people using things that they shouldn't and shipping software, or shipping things they thought were solid and are now gone... and we're dealing with cryptography here, so when things blow up, really, really bad things happen.

I don't think we have consensus on the following:

1. Is there going to be a did-core context and a did-core-registry context? 
2. How is the did-core context versioned, can it be cached, how can it be changed?
3. How is the did-core-registry context versioned, can it be cached, how can it be changed, if it can be changed, how can the versions be changed w/o causing headaches for implementers?
4. What constitutes stable releases for each of these items? Who decides? 

We need to understand all of these things before pulling in a file like this... and I don't want to do that right now (because we'll have to raise it to the group), I just want to pull in the other good stuff in this PR.

To give you a concrete example of how this file is going to blow up in Digital Bazaar's face -- you have rewritten the URL for publicKey - it was "https://w3id.org/security#publicKey" and now it's "https://w3c.github.io/did-core-registry/#publicKey"... which means anyone that pulls in this context and attempts to interop with Veres One, or use any of the Linked Data Signature tooling to interop, will see tons of exceptions being thrown... because the tooling will detect that the "publicKey" you are introducing here is totally different from the semantics of the "publicKey" that tooling has been using for 8+ years.

Surgery of this magnitude needs a concrete plan in place before we do it to ensure we don't break everything... and this context file has a strong possibility of making things even worse than they are now.

Here's what I think we should do as next steps:

1. Create an /ns/did/experimental/v2020.2 context and an /ns/did/registry/v2020.2 context (handwaving on names).
2. Update the /ns/did/experimental/v2020.2 to include all the latest DID Core stuff.
3. Update the /ns/did/registry/v2020.2 to include all the kitchen sink stuff like publicKeyHex.
4. Keep the source file for /ns/did/experimental/v2020.2 in did-core, because it's the did-core spec that defines what can go in there.
5. Keep the source file for /ns/did/registry/v2020.2 in this repo.

This still doesn't deal with the Linked Data Security stuff, but the plan there could be:

1. Move all files over to CCG, get redirect working to point there instead of the old domains.
2. Update all the human readable documentation.
3. Refer to this stuff from the context stuff above.

Fundamentally, we need plans like that before we can even consider doing something like this context file change... and the plan has to ensure that it doesn't blow up the ecosystem or fragment it further, etc.

Reply:
Thanks, this comment is very helpful. 

Regarding multiple contexts, my goal is to get a single jsonld context and human readable definition checked in somewhere so that we can point w3id here... and stop have so many separate repo's to update.

If this repo does not contain the did core jsonld context, along with human readable documentation for it, I can't really support this registry approach at all. 

the redirection services are pointing to repos that I can't fix. 

we can't point them here until this repo works... we can't make this a circular blocker for fixing things, the current interop tests are almost all failing... 

- https://identity.foundation/context/test-report.html

What I am proposing is the shortest path to getting to results with mostly passing tests, without needing to override security services as I currently need to do...

I'm not going to use w3id.org for anything that does not actually have documentation for it ...in other words the breaking difference is intentional, and can only be fixed by changing what w3id.org points to, until then, we should be explicit about where vocabulary is defined (github).

I agree that this is some big surgery, but its critically needed, and I am volunteering to do it.

What we have today in DIF is a registry with JSON-LD and JSON Schema and Human Readable documentation, in 1 repo.

If that repo were this repo, we could point w3id.org / w3c / here and be done.

the major differences between that repo and this repo are that its built with vanilla markdown and not respec, and that it has definitions for every property referenced by the did core contexts... even if the documentation just says deprecated / no documentation available.

I'm suggesting we convert the DIF repo to respec, move it here, and make it programmatically generated, before we attempt to fix w3id.org redirection services, which I am strongly opposed to generally, but which we can easily do as the last step, which will recover interoperability with the existing v1  / lds toolchains. 

I don't want people to think they can use this stuff before they can.

we should consider marking the registry unusable until it has no breaking issues like the ones related to security definitions.... 

The highest priority blocking issue is:

Which single repo contains the did core jsonld contexts (latest and versions), json schema and human readable documentation for both... if it is not this registry, I object to the proposed structure.

Second highest priority is how are we going to give these things nice URLs / not break existing stuff:

We should not use redirection services until the github hosted side of this stuff is shown to be working... at which point we should point w3id.org and w3c to whichever repo holds the machine and human readable registry information.

Reply:
The end of your comment is fine and I agree with it in general (I think)... but what you say at the top of the comment can be read in such a way that it contradicts what you say at the end. 

@OR13 wrote:
> Regarding multiple contexts, my goal is to get a single jsonld context and human readable definition checked in somewhere so that we can point w3id here... and stop have so many separate repo's to update.

"multiple contexts" could be read as you thinking that the security/v1, security/v2, did/v1, and did-core-registry/v1 JSON-LD contexts should all be combined into one context. There are variations in there that would be a mistake. 

For the rest of this comment, I'm assuming you mean:

* We leave the security/v1 and security/v2 contexts alone and will deal with them later.
* We move the did/v1 context into this repo (using did/v0.11 as the base with minor fix ups).
* We create a new did/registry/v1 context (or something like that) and put it into this repo.
* We publish all did/v1 and did/registry/v1 documentation in this repo.

If that's what you're saying, I'm on board.

> If this repo does not contain the did core jsonld context, along with human readable documentation for it, I can't really support this registry approach at all.

That's fine, we can do that, but we do need to get agreement from the group to do that because it's a structural change. I suggest that is done in a separate PR.

> Which single repo contains the did core jsonld contexts (latest and versions), json schema and human readable documentation for both... if it is not this registry, I object to the proposed structure.

As long as we don't conflate the DID Core context w/ the DID Registry context, I'm fine. If you think those two things are the same thing, then you're possibly not thinking about the security issues that are caused by a constantly changing base context as entries are added to the DID Core Registry.

> We should not use redirection services until the github hosted side of this stuff is shown to be working... at which point we should point w3id.org and w3c to whichever repo holds the machine and human readable registry information.

Yes, agreed... but that's not the point I was making... the point I was making is that you were fundamentally breaking the semantic URLs that are used. Now that you've stated that you don't want any of this to work, then just use: "urn:BROKEN_ON_PURPOSE" as a placeholder instead of making it seem like this context is supposed to work. So change this:

> "sec": "https://w3c.github.io/did-core-registry/#",

to this:

> "sec": "urn:BROKEN_ON_PURPOSE:security#",

... or just do all this in a separate PR, so I can merge this one more quickly. :)

Reply:
Just finished reviewing this PR... it's ready to go, pull the context out of the PR and put it in a separate PR so I can merge.

Reply:
I have remove the context, we can tackle in subsequent PR.